design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+1. Custom error handling. You can get any error you wish for, handle it properly, and send it back.
+1. Versatility. Script can be customized later if needed to handle more issues in this area.
+1. Simple implementation
+
+This solution is implemented in [this PR](https://github.com/cyberark/conjur/pull/1789).
+
+#### Conclusion
+
+Option 2 is off the table due to security concerns.
+
+Option 3 is off the table due to redundant complexity and backwards-compatibiliy issues.
+
+Option 4 has the most benefits of the options and is better than option 1.
+
+<u>Decision</u>: Option 4
+
+### Design
+[//]: # "Add any diagrams, charts and explanations about the design aspect of the solution. Elaborate also about the expected user experience for the feature"
+The authentication process is depicted in the following diagram:
+![Authn-k8s authenti](authn-k8s-process.png)
+
+Note that the cert injection happens in a separate thread.
+
+### Backwards compatibility
+[//]: # "Address how you are going to handle backwards compatibility, if necessary"
+We don't have backwards-compatible issues with this change as also before this change the server didn't wait for the cert to be injected. The server closed the socket to kubectl before waiting for the cert to be injected. Furthermore, the authn-client just verifies that the response code is less than 300 so changing 200 to 202 doesn't affect it.
+
+### Performance
+[//]: # "Elaborate on whether your solution will affect the product's performance, and how"
+The performance of the "inject-client-cert" request will not be affected as the Websocket is still being closed immediately after we inject the cert.
+
+### Security
+
+[//]: # "Are there any security issues with your solution? Even if you mentioned them somewhere in the doc it may be convenient for the security architect review to have them centralized here"
+
+1. Is it ok to pass the base64 encoded cert in the URL? Is it less secure than sending it in the STDIN?
+   1. The answer is yes and thus we are not going with option 2.
+
+## Affected Components
+
+[//]: # "Address the components that will be affected by your solution [Conjur, DAP, clients, integrations, etc.]"
+- Conjur
+- DAP
+- conjur-authn-k8s-client
+- seed-fetcher
+- secrets-provider-for-k8s
+- secretless-broker
+
+## Cross-team dependencies
+[//]: # "Elaborate about your solution's requirements or needs from other teams"
+We would like to get a review on the solution from @cyberark/conjur-core-team  and @cyberark/community-and-integrations-team. There are no requirements from  the Infra team.
+
+## Test Plan
+[//]: # "Fill in the table below to depict the tests that should run to validate your solution"
+[//]: # "You can use this tool to generate a table - https://www.tablesgenerator.com/markdown_tables#"
+
+The current integration tests of the "inject-client-cert" request will be sufficient for this feature. we will just change them so they verify the response is 202 instead of 200.
```

Comment:
Moti, suggested we invest in performance tests because we don't really know the numbers and don't want to worsen the performance experience with the changes in the PR